runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -1270,9 +1379,20 @@ TR_J9InlinerPolicy::createUnsafePutWithOffset(TR::ResolvedMethodSymbol *calleeSy
       comp()->getDebug()->print(comp()->getOutFile(), directAccessTreeTop);
       }
 
-   TR::TreeTop* arrayDirectAccessTreeTop = conversionNeeded
-      ? genDirectAccessCodeForUnsafeGetPut(unsafeNodeWithConversion, conversionNeeded, false)
-      : NULL;
+   // We need to generate an arrayDirectAccessBlock AND a directAccessBlock in the following cases:
+   // 1.) arrayTestNeeded == true: arrayDirectAccessBlock used for array access,
+   //                               directAccessBlock used for non-array/non-class object access
+   // 2.) !arrayTestNeeded && arrayBlockNeeded: arrayDirectAccessBlock used for array access,
+   //                                            directAccessBlock used if object is NULL
+   // Otherwise, only the directAccessBlock is needed.
+   TR::TreeTop* arrayDirectAccessTreeTop;
+
+   if (conversionNeeded && (arrayTestNeeded|| arrayBlockNeeded))
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(unsafeNodeWithConversion, conversionNeeded, false);
+   else if (arrayTestNeeded || arrayBlockNeeded)
+      arrayDirectAccessTreeTop = genDirectAccessCodeForUnsafeGetPut(unsafeNode, false, false);
```

Comment:
I just realized that in this case, if `javaLangClass == NULL`, but `conversionNeeded` is false and `TR::Compiler->om.isOffHeapAllocationEnabled()` is false, the trees that are generated for `arrayDirectAccessTreeTop` are identical to those generated for `directAccessTreetop` above.

I can see that that's necessary for the off-heap allocation case, as you need to tailor the IL under `arrayDirectAccessTreeTop` for the off-heap array access below, but otherwise, this results in duplicated blocks.  The duplicated block can be seen in the [log you generated earlier](https://github.com/user-attachments/files/16133509/gencon_int_unknown_javalangclassNULL.log) for the `javaLangClass == NULL` case &mdash; block_9 and block_8 are the same (except for a `goto` operation):

```
n105n     BBStart <block_12>                                                                  [    0x7c44f30a1090] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n104n     ificmpne --> block_8 BBStart at n86n ()                                             [    0x7c44f30a1040] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=2 flg=0x20
n101n       iand                                                                              [    0x7c44f30a0f50] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=2
n99n          l2i                                                                             [    0x7c44f30a0eb0] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=1
n98n            lloadi  <isClassDepthAndFlags>[#317  Shadow +24] [flags 0x603 0x0 ]           [    0x7c44f30a0e60] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=1
n95n              aloadi  <vft-symbol>[#320  Shadow] [flags 0x18607 0x0 ]                     [    0x7c44f30a0d70] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=1
n94n                aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                         [    0x7c44f30a0d20] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n100n         iconst 0x10000 (X!=0 X>=0 )                                                     [    0x7c44f30a0f00] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x104
n103n       iconst 0                                                                          [    0x7c44f30a0ff0] bci=[-1,0,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n106n     BBEnd </block_12> =====                                                             [    0x7c44f30a10e0] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=0

n96n      BBStart <block_11>                                                                  [    0x7c44f30a0dc0] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n83n      iflcmpne --> block_10 BBStart at n91n ()                                            [    0x7c44f30a09b0] bci=[-1,4,33] rc=0 vc=0 vn=- li=- udi=- nc=2 flg=0x20
n82n        land                                                                              [    0x7c44f30a0960] bci=[-1,4,33] rc=1 vc=0 vn=- li=- udi=- nc=2
n47n          lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                               [    0x7c44f30054c0] bci=[-1,4,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n81n          lconst 1 (highWordZero X!=0 X>=0 )                                              [    0x7c44f30a0910] bci=[-1,0,33] rc=2 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
n81n        ==>lconst 1
n97n      BBEnd </block_11> =====                                                             [    0x7c44f30a0e10] bci=[-1,3,33] rc=0 vc=0 vn=- li=- udi=- nc=0

n89n      BBStart <block_9>                                                                   [    0x7c44f30a0b90] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n53n      istore  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ]                                  [    0x7c44f30a0050] bci=[-1,5,33] rc=0 vc=14 vn=- li=- udi=- nc=1
n54n        iloadi  <unsafe shadow sym>[#420  Shadow] [flags 0x80000603 0x100 ] ()            [    0x7c44f30a00a0] bci=[-1,5,33] rc=1 vc=15 vn=- li=- udi=- nc=1 flg=0x20
n55n          aladd                                                                           [    0x7c44f30a00f0] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=2
n56n            aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                             [    0x7c44f30a0140] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n58n            lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                             [    0x7c44f30a01e0] bci=[-1,4,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n90n      BBEnd </block_9> =====                                                              [    0x7c44f30a0be0] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0

n84n      BBStart <block_7>                                                                   [    0x7c44f30a0a00] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n8n       ireturn                                                                             [    0x7c44f3004890] bci=[-1,8,33] rc=0 vc=14 vn=- li=- udi=- nc=1
n6n         iload  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ] ()                              [    0x7c44f30047f0] bci=[-1,5,33] rc=1 vc=15 vn=- li=- udi=- nc=0 flg=0x20
n2n       BBEnd </block_7> =====                                                              [    0x7c44f30046b0] bci=[-1,8,33] rc=0 vc=14 vn=- li=- udi=- nc=0

n86n      BBStart <block_8>                                                                   [    0x7c44f30a0aa0] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n60n      istore  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ]                                  [    0x7c44f30a0280] bci=[-1,5,33] rc=0 vc=14 vn=- li=- udi=- nc=1
n61n        iloadi  <unsafe shadow sym>[#420  Shadow] [flags 0x80000603 0x100 ] ()            [    0x7c44f30a02d0] bci=[-1,5,33] rc=1 vc=15 vn=- li=- udi=- nc=1 flg=0x20
n62n          aladd                                                                           [    0x7c44f30a0320] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=2
n63n            aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                             [    0x7c44f30a0370] bci=[-1,3,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n65n            lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                             [    0x7c44f30a0410] bci=[-1,4,33] rc=1 vc=0 vn=- li=- udi=- nc=0
n88n      goto --> block_7 BBStart at n84n                                                    [    0x7c44f30a0b40] bci=[-1,5,33] rc=0 vc=0 vn=- li=- udi=- nc=0
n87n      BBEnd </block_8> =====                             
```

This might not be terribly important, as I hope the `javaLangClass == NULL` condition doesn't occur frequently, but it might be worth thinking about.

Reply:
Thinking about it a little more, I think this could be handled with `arrayTestNeeded` being set to true, but `arrayBlockNeeded` set to false.  `genCodeForUnsafeGetPut` has had to deal with that scenario in the past.

Reply:
Thank you for catching that! I've modified my changes to handle this situation as a special case, since as you said, it's not something we expected to come up too much. In the situation where gencon is being used and `conversionNeeded` is false, both sides of the array test diamond will lead to the `directAccessBlock`, so we can just skip the array test altogether. This leaves just the NULL and lowtag tests, like so:
```
n1n       BBStart <block_2>                                                                   [    0x71c90ec14660] bci=[-1,0,36] rc=0 vc=14 vn=- li=- udi=- nc=0
n42n      NULLCHK on n3n [#32]                                                                [    0x71c90ec15330] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=1
n41n        PassThrough                                                                       [    0x71c90ec152e0] bci=[-1,5,36] rc=1 vc=0 vn=- li=- udi=- nc=1
n3n           aload  TestUnsafe.unsafe Lsun/misc/Unsafe;[#405  Static] [flags 0x307 0x0 ]     [    0x71c90ec14700] bci=[-1,0,36] rc=1 vc=14 vn=- li=- udi=- nc=0
n44n      astore  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                                  [    0x71c90ec153d0] bci=[-1,3,36] rc=0 vc=0 vn=- li=- udi=- nc=1
n4n         aload  <parm 0 Ljava/lang/Object;>[#403  Parm] [flags 0x40000107 0x0 ]            [    0x71c90ec14750] bci=[-1,3,36] rc=1 vc=14 vn=- li=- udi=- nc=0
n46n      lstore  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                                  [    0x71c90ec15470] bci=[-1,4,36] rc=0 vc=0 vn=- li=- udi=- nc=1
n5n         lload  <parm 1 J>[#404  Parm] [flags 0x40000104 0x0 ]                             [    0x71c90ec147a0] bci=[-1,4,36] rc=1 vc=14 vn=- li=- udi=- nc=0
n73n      ifacmpeq --> block_9 BBStart at n82n ()                                             [    0x71c90ecb0690] bci=[-1,3,36] rc=0 vc=0 vn=- li=- udi=- nc=2 flg=0x20
n71n        aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                                 [    0x71c90ecb05f0] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n72n        aconst NULL (X==0 X>=0 X<=0 )                                                     [    0x71c90ecb0640] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x302
n78n      BBEnd </block_2> =====                                                              [    0x71c90ecb0820] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0

n86n      BBStart <block_10>                                                                  [    0x71c90ecb0aa0] bci=[-1,3,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n76n      iflcmpeq --> block_8 BBStart at n79n ()                                             [    0x71c90ecb0780] bci=[-1,4,36] rc=0 vc=0 vn=- li=- udi=- nc=2 flg=0x20
n75n        land                                                                              [    0x71c90ecb0730] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=2
n47n          lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                               [    0x71c90ec154c0] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n74n          lconst 1 (highWordZero X!=0 X>=0 )                                              [    0x71c90ecb06e0] bci=[-1,0,36] rc=2 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
n74n        ==>lconst 1
n87n      BBEnd </block_10> =====                                                             [    0x71c90ecb0af0] bci=[-1,3,36] rc=0 vc=0 vn=- li=- udi=- nc=0

n82n      BBStart <block_9>                                                                   [    0x71c90ecb0960] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n53n      istore  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ]                                  [    0x71c90ecb0050] bci=[-1,5,36] rc=0 vc=14 vn=- li=- udi=- nc=1
n54n        iloadi  <unsafe shadow sym>[#420  Shadow] [flags 0x80000603 0x100 ] ()            [    0x71c90ecb00a0] bci=[-1,5,36] rc=1 vc=15 vn=- li=- udi=- nc=1 flg=0x20
n55n          aladd                                                                           [    0x71c90ecb00f0] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=2
n56n            aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                             [    0x71c90ecb0140] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n58n            lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                             [    0x71c90ecb01e0] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n83n      BBEnd </block_9> =====                                                              [    0x71c90ecb09b0] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0

n77n      BBStart <block_7>                                                                   [    0x71c90ecb07d0] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n8n       ireturn                                                                             [    0x71c90ec14890] bci=[-1,8,36] rc=0 vc=14 vn=- li=- udi=- nc=1
n6n         iload  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ] ()                              [    0x71c90ec147f0] bci=[-1,5,36] rc=1 vc=15 vn=- li=- udi=- nc=0 flg=0x20
n2n       BBEnd </block_7> =====                                                              [    0x71c90ec146b0] bci=[-1,8,36] rc=0 vc=14 vn=- li=- udi=- nc=0

n79n      BBStart <block_8> (freq 1) (cold)                                                   [    0x71c90ecb0870] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n62n      istore  <temp slot 6>[#421  Auto] [flags 0x3 0x0 ]                                  [    0x71c90ecb0320] bci=[-1,5,36] rc=0 vc=14 vn=- li=- udi=- nc=1
n63n        iloadi  <unsafe shadow sym>[#422  Shadow] [flags 0x80000603 0x100 ] ()            [    0x71c90ecb0370] bci=[-1,5,36] rc=1 vc=15 vn=- li=- udi=- nc=1 flg=0x20
n64n          aladd                                                                           [    0x71c90ecb03c0] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=2
n65n            aloadi  <ramStaticsFromClass>[#313  Shadow +72] [flags 0x10607 0x0 ]          [    0x71c90ecb0410] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=1
n66n              aloadi  <classFromJavaLangClass>[#311  Shadow +8] [flags 0x10607 0x0 ]      [    0x71c90ecb0460] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=1
n67n                aload  <temp slot 4>[#418  Auto] [flags 0x7 0x0 ]                         [    0x71c90ecb04b0] bci=[-1,3,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n68n            land                                                                          [    0x71c90ecb0500] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=2
n69n              lload  <temp slot 5>[#419  Auto] [flags 0x4 0x0 ]                           [    0x71c90ecb0550] bci=[-1,4,36] rc=1 vc=0 vn=- li=- udi=- nc=0
n70n              lconst -4 (X!=0 X<=0 )                                                      [    0x71c90ecb05a0] bci=[-1,5,36] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x204
n81n      goto --> block_7 BBStart at n77n                                                    [    0x71c90ecb0910] bci=[-1,5,36] rc=0 vc=0 vn=- li=- udi=- nc=0
n80n      BBEnd </block_8> (cold) =====  
```
(The full trace log can be viewed [here](https://github.com/user-attachments/files/16240212/gencon_int_unknown_javaLangClassNULL.log).)

In adding this additional case, I also found that I was able to simplify the logic surrounding `typeTestsNeeded`, `arrayBlockNeeded`, and `arrayTestNeeded` as well: namely, I was able to get rid of `arrayTestNeeded` entirely, since it is equivalent to `typeTestsNeeded && arrayBlockNeeded`. So with that being the case, the five cases that `genCodeForUnsafeGetPut()` needs to handle are as follows (with case 3 being the one that I added just now):
```
   // There are 4 possible cases determining which checks and blocks should be added to the generated IL trees:
   // 1.) typeTestsNeeded && arrayBlockNeeded
   //     (i.e.: object type is unknown at compile time AND (offheap OR data element size < 4 (byte, short)))
   //     - tests: NULL test, Array test, Lowtag test
   //     - blocks: arrayDirectAccess, directAccess, indirectAccess
   // 2.) typeTestsNeeded && !arrayBlockNeeded && javaLangClass != NULL
   //     (i.e.: object type is unknown at compile time AND gencon AND data element size >= 4 (int, long) AND able to java/lang/Class)
   //     - tests: Lowtag test, NULL test, Class test
   //     - blocks: directAccess, indirectAccess
   // 3.) typeTestsNeeded && !arrayBlockNeeded && javaLangClass == NULL
   //     (i.e.: object type is unknown at compile time AND gencon AND data element size >= 4 (int, long) AND unable to java/lang/Class)
   //     - tests: NULL test, Lowtag test
   //     - blocks: directAccess, indirectAccess
   // 4.) !typeTestsNeeded && arrayBlockNeeded
   //     (i.e.: object is known to be array at compile time AND (offheap OR data element size < 4 (byte, short)))
   //     - tests: NULL test
   //     - blocks: arrayDirectAccess, directAccess
   // 5.) !typeTestsNeeded && !arrayBlockNeeded
   //     (i.e.: object is known to be non-array at compile time)
   //     - tests: none
   //     - blocks: directAccess
```
So now `javaLangClass == NULL` is separated out from the other conditions and only has an effect if `typeTestsNeeded == true` and `arrayBlockNeeded == false` (cases 2 and 3), which should cover the scenario that you mentioned above.

Reply:
> In the situation where gencon is being used and `conversionNeeded` is false, both sides of the array test diamond will lead to the `directAccessBlock`, so we can just skip the array test altogether. This leaves just the NULL and lowtag tests. . . .

I'm not sure that it's permissible to omit the array test in this case.  I haven't confirmed I'm correct about this, but I think it's possible that code could be writing or reading across array elements, even if conversion is not needed - for instance, I think it's possible that the caller could give an offset of 17 for an int array in a call to `Unsafe.putInt`, in order to write beginning at the last byte of one element and into the first three bytes of the next element.

Previously if `conversionNeeded` is false, the IL that was generated &mdash; for both cases where `javaLangClass` is null or is not null &mdash; allowed for the possibility that the low bit of the offset is set even if the location being stored to/loaded from is some Java object and not a static field.  The only way I can think that might be possible is with unaligned access across array elements - although `Unsafe` also has `Unaligned` methods that seem like they ought to be used in such situations.

Does anyone with more knowledge of how `Unsafe` works or is used understand why it needs to allow for the possibility that the low bit of the offset might be set for `{get|put}{Int|Long}` even if the target is not a static field?  @vijaysun-omr, @jdmpapin?

Reply:
`get` and `put` for smaller integral types need to handle the case where a naturally aligned array element is not aligned to a 4-byte boundary. Is it possible that the logic to do so was needlessly applied for int and long as well?

The javadoc in `Unsafe` heavily restricts the valid offsets that can be passed. In the case of an array, the offset must be the array type's base offset plus a multiple of its index scale. Additionally, the array component type must be as expected by the method, e.g. for `getInt()` it must be `int`. Together these mean that anything passing a misaligned offset is violating the preconditions of the method. So *hopefully* nothing does that

Starting in Java 11 (though maybe introduced in 9 or 10) we have `Unsafe.getIntUnaligned()` and so on, which are there specifically to allow for potentially misaligned accesses. There is currently no special handling for those in the JIT compiler, but #19668 is open to treat them more or less just like their aligned counterparts. If the trees we generate for `getInt()` don't allow misaligned offsets, or if they stop allowing them, then the trees for `getIntUnaligned()` will need to differ in that respect (and similarly for `putInt()` and `get`/`put` of other types)

Reply:
i am wondering how directByteBuffer is accessed currently, whether it is through Unsafe APIs (obviously conversion must be prohibited, otherwise you can access over the buffer limit)?
if not, we might be able to get rid of the low-tagging shenanigans (J9 implementation-specific?).

Reply:
> Is it possible that the logic to do so was needlessly applied for int and long as well?

That's what I was wondering.  I'd been assuming these past few months that we were intentionally allowing for the possibility of unaligned access for int and long, but now I'm unsure about whether that was inadvertent.  I spent some time digging into the history, but it wasn't readily obvious whether that possibility was being explicitly allowed for.

However, I've just looked at the JVM's underlying implementation of `Unsafe.getLong()` in [`get64`](https://github.com/eclipse-openj9/openj9/blob/0e944e75c874f243f8f14c402ea8f268bf5e78b7/runtime/oti/UnsafeAPI.hpp#L262-L294).  It checks whether the object is an array before testing the least significant bit of the offset, and it's allowing for the possibility that an array access is misaligned.  So I think that the JIT's behaviour really was intended.

@tajila, am I understanding correctly?

Reply:
@hzongaro Yes, we do allow unaligned access with get/put* on arrays such that a read/load can cross an element boundary. Historically (pre-JDK9), this was always true. The VM implementation will always support this because we will have one version for pre and post JDK9. 

That being said,  recently tried this with an RI JDK17 build and it still allows unaligned access on an array.

> The javadoc in Unsafe heavily restricts the valid offsets that can be passed.

@jdmpapin where did you find this? it wasnt immediately clear to me.

Reply:
Unsafe.java in the extension repos (sun/misc in JDK8, jdk/internal/misc in JDK11+). Look at the javadoc comment for `getInt()`, which is referenced from `putInt()` and other `get`/`put` methods

Reply:
Oh, I guess we have our own Unsafe.java for 9+. But here's the [relevant part of the javadoc comment](https://github.com/ibmruntimes/openj9-openjdk-jdk8/blob/e642f8b86c560be492fa2d3c5a896d6ac5340021/jdk/src/share/classes/sun/misc/Unsafe.java#L107-L131) in Java 8

Reply:
Thanks, I was looking at the OpenJ9 docs. I wouldn't characterize "results are undefined" as a restrication, it is Unsafe after all.

Regardless, based on experimentation it seems that we still have to support unaligned access in get/putInt/Long at least on x86. There is no verification of parameters for Unsafe APIs so I suspect on some platforms it may just sigbus fault.

Reply:
I interpreted "results are undefined" to mean that we hit undefined behaviour, i.e. that the implementation is theoretically free to assume that the arguments satisfy the stated conditions. I think generally for escape hatches like `Unsafe`, the possibility of arbitrary consequences for misuse (i.e. UB) is fundamentally what makes them "unsafe." If the comment just meant to say that the result/return value is unpredictable then it could have said that instead. AFAICT the comment is worded very carefully so that it applies to all of the regular `get`/`put` operations, which refer to it. Especially in the context of a `put`, I don't know how else to interpret "results are undefined"

But even on that interpretation, since we've been allowing a particular use, there might be code that relies on it. I guess your experimentation has shown that something is actually relying on it?

Reply:
Your interpretation of that comment sounds correct, but this would not be the first time an API defines certain pre-conditions yet the RI goes to great lengths to suppport cases outside those condition. 

Generally speaking, we treat "undefined behaviour" as spec'ed by the RI. Its hard to know who out there is relying on what behaviour. Given that unsafe is quite pervasive, I think the safest option is to aim for parity.

Reply:
Since unaligned array accesses are a possibility that I hadn't considered, I'll add back that array check for now. I'm not sure if there's another, more efficient way to account for this situation, but in the interest of getting these changes in soon, I think taking the safe approach here, and then adding any further improvements (if necessary) in a future contribution is probably the best course of action.

Reply:
With this change, the trees for the gencon/`javaLangClass == NULL` case can be viewed here: 
[gencon_int_unknown_javaLangClassNULL.log](https://github.com/user-attachments/files/16287454/gencon_int_unknown_javaLangClassNULL.log)